addons/partner_autocomplete/models/res_partner.py

```diff
@@ -147,7 +148,20 @@ def _is_company_in_europe(self, country_code):
     def _is_vat_syncable(self, vat):
         vat_country_code = vat[:2]
         partner_country_code = self.country_id.code if self.country_id else ''
-        return self._is_company_in_europe(vat_country_code) and (partner_country_code == vat_country_code or not partner_country_code)
+        return self._is_company_in_europe(vat_country_code) and (partner_country_code == vat_country_code or not partner_country_code) or self.check_gst_in(vat)
+
+    def check_gst_in(self, vat):
+        # reference from https://www.gstzen.in/a/format-of-a-gst-number-gstin.html
+        if vat and len(vat) == 15:
+            all_gstin_re = [
+                r'[0-9]{2}[a-zA-Z]{5}[0-9]{4}[a-zA-Z]{1}[1-9A-Za-z]{1}[Zz1-9A-Ja-j]{1}[0-9a-zA-Z]{1}',  # Normal, Composite, Casual GSTIN
+                r'[0-9]{4}[A-Z]{3}[0-9]{5}[UO]{1}[N][A-Z0-9]{1}',  # UN/ON Body GSTIN
+                r'[0-9]{4}[a-zA-Z]{3}[0-9]{5}[N][R][0-9a-zA-Z]{1}',  # NRI GSTIN
+                r'[0-9]{2}[a-zA-Z]{4}[a-zA-Z0-9]{1}[0-9]{4}[a-zA-Z]{1}[1-9A-Za-z]{1}[DK]{1}[0-9a-zA-Z]{1}',  # TDS GSTIN
+                r'[0-9]{2}[a-zA-Z]{5}[0-9]{4}[a-zA-Z]{1}[1-9A-Za-z]{1}[C]{1}[0-9a-zA-Z]{1}'  # TCS GSTIN
```

Comment:
These regexes can be simplified a bit:
- `[0-9]` can be replaced by `\d`.
- Set of characters containing a single character, such as `[N]`, shouldn't be set of characters but the character itself.
- Occurrences of `{1}` can be removed, it's the default behavior to have one copy of the thing.